pkg/destroy/aws/aws.go

```diff
@@ -641,35 +646,110 @@ func deleteEC2Instance(ec2Client *ec2.EC2, iamClient *iam.IAM, id string, logger
 
 	for _, reservation := range response.Reservations {
 		for _, instance := range reservation.Instances {
-			// Skip 'terminated' instances since they take a while to really get cleaned up
-			if *instance.State.Name == "terminated" {
-				continue
-			}
-			if instance.IamInstanceProfile != nil {
-				parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
-				if err != nil {
-					return errors.Wrap(err, "parse ARN for IAM instance profile")
-				}
-
-				err = deleteIAMInstanceProfile(iamClient, parsed, logger)
-				if err != nil {
-					return errors.Wrapf(err, "deleting %s", parsed.String())
-				}
-			}
-
-			_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
-				InstanceIds: []*string{instance.InstanceId},
-			})
+			err = deleteEC2InstanceByInstance(ec2Client, iamClient, instance, logger)
 			if err != nil {
 				return err
 			}
+		}
+	}
+	return nil
+}
+
+func deleteEC2InstanceByInstance(ec2Client *ec2.EC2, iamClient *iam.IAM, instance *ec2.Instance, logger logrus.FieldLogger) error {
+	// Skip 'terminated' instances since they take a while to really get cleaned up
+	if instance.State == nil || *instance.State.Name == "terminated" {
+		return nil
+	}
 
-			logger.Info("Deleted")
+	if instance.IamInstanceProfile != nil {
+		parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
+		if err != nil {
+			return errors.Wrap(err, "parse ARN for IAM instance profile")
+		}
+
+		err = deleteIAMInstanceProfile(iamClient, parsed, logger)
+		if err != nil {
+			return errors.Wrapf(err, "deleting %s", parsed.String())
 		}
 	}
+
+	if *instance.State.Name == "shutting-down" {
+		return nil
+	}
+
+	_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
+		InstanceIds: []*string{instance.InstanceId},
+	})
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Terminating")
```

Comment:
Hmm, don't we INFO only `deleted` stuff?

Reply:
> Hmm, don't we INFO only `deleted` stuff?

We used to [info-log `Deleted` when a termination request was accepted][1].  This log line preserves that approach with a better string, because you can't actually delete instances, and we aren't waiting around to see AWS eventually reap the terminated instances.

[1]: https://github.com/openshift/installer/blob/58205006353c4be4a1f1a8fce95136020d6ede95/pkg/destroy/aws/aws.go#L660-L667

Reply:
but when do we log `Terminated`? You capture the terminated when `ByTags`, but if we get an instance by deleteARN you'll only see `Terminating`.. that's not ideal.

>  This log line preserves that approach with a better string, because you can't actually delete instances

but to be fair, the for all intents and purposes the instance will be delete at some time.. after the termination request was sent, so the `Deleted` wasn't wrong.

Reply:
>  This log line preserves that approach with a better string, because you can't actually delete instances, and we aren't waiting around to see AWS eventually reap the terminated instances.

also in that case the `deleteEC2InstanceByInstance` isn't appropriate too..

Reply:
> also in that case the `deleteEC2InstanceByInstance` isn't appropriate too..

`s/deleteEC2Instance/terminateEC2Instance` across the file with 6af061afd -> 54cfed55a.

> but when do we log `Terminated`?

Never, with the `deleteARN` chain.  I'm fine with that, because tracking these outside of `terminateEC2InstancesByTags` just so we can report the terminated state doesn't seem like it's worth the trouble to me.  Do you think I need to work that in?

> but to be fair, the for all intents and purposes the instance will be delete at some time...

Which is why I'm fine leaving things at `Terminating` when we don't have a reason to wait out the termination (like we do in `terminateEC2InstancesByTags` but which we don't with the `deleteARN` tree).

> ... after the termination request was sent, so the `Deleted` wasn't wrong.

I still think `Terminating` is better than `Deleted` for "my `TerminateInstances` call just went through".  But yeah, it's splitting a bit of a hair.  If you feel like I'm being overly picky, I can shift things back to `Deleted`.